### DISCUSSION TITLE
Need help in combine the Imagemagick Commands

### DISCUSSION THREAD

**Discussion by gmek5476** [ImageMagick/ImageMagick#6326] [Development]
### ImageMagick version

7.1.1-7

### Operating system, version and so on

Windows

### Description

I have to create a gradient for an image iam using below commands . in two steps in the First step creating a gradient overlay image and applying that layoverimage to our source image. I Need help in combinign these two commands and make it in one execetion.

1.magick -size 100x200 -define  gradient:direction=east gradient:black-none C:\Im-Test\Output\gradient\gr_12_baseGrad.png

2.convert C:\Im-Test\Output\Ketchup-S12.jpg C:\Im-Test\Output\gradient\gr_12_baseGrad.png -gravity West -composite C:\Im-Test\Output\gradient\G_Final_123.jpg

Any help is appreciated.

**Comment by fmw42** [comment]
Suggest you don't use convert in IM 7. Use Magick.  Does this do what you want?

```
magick C:\Im-Test\Output\Ketchup-S12.jpg ^
( -size 100x200 -define gradient:direction=east gradient:black-none ) ^
-gravity West -composite C:\Im-Test\Output\gradient\G_Final_123.jpg
```

Be sure there is a space on either side of each parenthesis. The above is a 3-line single command. The ^ indicate new line after it. So copy and past as is in 3 lines.  You may remove the ^s but then make sure the command is on one single line of text.

See https://imagemagick.org/Usage/basics/#parenthesis

**Reply by gmek5476** [reply]
Thanks Fred this Worked OUT. And one more question how can we add opacity and gradient percentage to the existing command.
Thanks in Advance

**Reply by fmw42** [reply]
Please clarify.

1) Opacity in the output or somewhere else?

2) What does gradient percent mean -- what should it control?

**Reply by gmek5476** [reply]
I refer opacity means the transperancy in the gradient and gradient percentage means the gradient extent .For example if i want gradient of and image with 80% darkness of one color and remaining 20% with another color.

**Reply by fmw42** [reply]
```
magick C:\Im-Test\Output\Ketchup-S12.jpg ^
( -size 100x200 -define gradient:direction=east gradient:"rgba(r,g,b,a )-rgba(r,g,b,a)" ^
-gravity West -composite C:\Im-Test\Output\gradient\G_Final_123.jpg
```

Where two different rgba(r,g,b,a) color values are used.    For example: 50% opacity red is rgba(255,0,0,0.5).  Or if you want gray colors,  25% opacity mid gray is graya(128,0.25). Or you can use values as %. For example mid gray with 25% opacity would be graya(50%,0.25).  See color definitions at https://imagemagick.org/script/color.php

**Reply by gmek5476** [reply]
Hi Fred ,Thanks for your time i need your help in one more case. To the previous command i want to apply cropping and resizing to the source image. Please find my two commands below. In the First command iam cropping the source image and resizing the result as the ht and width. In the second command iam applying the gradient. Could you please help me to combine these two instructions into one command.Appreciate your help and support.

1.convert C:\Im-Test\Input\Ketchup-S.jpg -crop 437x993+825+0 +repage -adaptive-resize "222X500^" -gravity center -crop 222x500+0+0 C:\Im-Test\Output\Ketchup-default_12456.jpg

2.( -size 100x200 -define gradient:direction=east gradient:"rgba(r,g,b,a )-rgba(r,g,b,a)" ^
-gravity West -composite C:\Im-Test\Output\gradient\G_Final_123.jpg

**Reply by fmw42** [reply]
Your second command does not use the first image. Also your second command had no convert and no input image.  Please correct your second command as it is, it is not a proper command.

**Reply by gmek5476** [reply]
Sorry here is my second command
convert "/mnt/shared/dev_mm/amd_repo/3/e/20/7c/e207c3467a391a4ded116c2ae318e3be10f9e57e/original/495_beyond-sausage-pizza-pesto-cream-cauliflower-crust-S.jpg[0]" -auto-orient -resize 300x200  "/mnt/shared/dev_mm/amd_repo//3/e/20/7c/e207c3467a391a4ded116c2ae318e3be10f9e57e/d2754896f17947af04bb715c5d7aa79f3cd52fb1.jpg"

**Reply by fmw42** [reply]
There is no connection between the images of your second command your images of the first command. So I cannot make them into one command line.

**Reply by gmek5476** [reply]
As of now iam running them independently in our java code, we are executing the first command and saving the response in a variable and using that variable as input to second command.

**Reply by fmw42** [reply]
Your second command now has no gradient.

Is the output of the first command being used as the input of the second command?  Please clarify and use common images with short names so it is easier for anyone to read your commands. 

If that is the case, then you just chain the command. 

```
convert input1 commands1 output1
convert output1 commands2 output2
```

becomes

`convert input1 commands1 commands2 output2`

**Reply by gmek5476** [reply]
Apologies for the confusion ,My clipboard cheated me and i have pasted wrong commands.Here is the commands which i supposed to sent.
1)convert "src.jpg" -crop 437x993+825+0 +repage -adaptive-resize "222X500^" -gravity center -crop 222x500+0+0 "inter-Output.jpg"
2)convert "inter-Output.jpg" ( -size 1000x348 -define  gradient:direction=east gradient:grey-none )  -gravity West -composite   "output.jpg"

From the above commands i want to use the output of first command as input of second command.
i have given short names tot he files for your convenience.
Onceagain iam sorry for the confusion.

**Reply by fmw42** [reply]
As I mentioned above, just chain the commands

```
convert "src.jpg" -crop 437x993+825+0 +repage -adaptive-resize "222X500^" -gravity center -crop 222x500+0+0 ^
( -size 1000x348 -define  gradient:direction=east gradient:grey-none )  -gravity West -composite   "output.jpg"
```

**Reply by gmek5476** [reply]
Hi Fred 
Could you please help with imagemagick command of achieving  target image as attached.
Ignore the text on target image.
<img width="226" alt="src" src="https://github.com/ImageMagick/ImageMagick/assets/126016862/37aa576a-c178-43a0-809d-e047ef2e4be4">
<img width="253" alt="target" src="https://github.com/ImageMagick/ImageMagick/assets/126016862/f1a29d8a-b4b1-4b3c-aacc-edcb087b194c">

**Reply by fmw42** [reply]
Please clarify what exactly you are trying to do with those two images.

**Reply by gmek5476** [reply]
The First image is Source image and Second image is that i want to achieve.
I tried to draw a polygon by using -draw in imagemagick,But i could understand how the polygon works to create and i cannot pass the co-ordinates properly du to lack of understandin -draw command. Really appreciate your help.

**Reply by fmw42** [reply]
What processing are you doing to the input.  The scale looks different.  Do you have a command for that part of your processing?

**Reply by gmek5476** [reply]
here is the command iam using 
convert <srcimage> ( +clone -fill Black -colorize 100 -fill White -draw "polygon 500,0 1500,0 1500,1000 0,1000" ) -alpha off -compose CopyOpacity -composite -trim +repage <target image>

It works partially where i couldn't change color and i want change the projection(means i have calculate coordinates based on angle).But i can't understand the how to compute co-ordinates based on angle degrees

**Reply by fmw42** [reply]
Your PDF is corrupt or is not readable by Ghostscript.  Imagemagick identify -verbose says:

```
identify: Postscript delegate failed `[ghostscript library 9.56] -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 '-sDEVICE=pngalpha' -dTextAlphaBits=4 -dGraphicsAlphaBits=4 '-r72x72' -dPrinted=false  '-sOutputFile=/tmp/magick-yHyodEHQ2O-cENxFiWZLJMinDAflslTJ%d' '-f/tmp/magick-jg5-_7h-Uo9T2rWcLepKsIU6npMOCDQT' '-f/tmp/magick-3tVV0RXjBUW_wq0k_uygOEmAWNkogRZp'': Error: /undefined in avatars.githubusercontent.com
Operand stack:

Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   1990   1   3   %oparray_pop   1989   1   3   %oparray_pop   1977   1   3   %oparray_pop   1833   1   3   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push
Dictionary stack:
   --dict:771/1123(ro)(G)--   --dict:0/20(G)--   --dict:75/200(L)--
Current allocation mode is local
GPL Ghostscript 9.56.1: Unrecoverable error, exit code 1
 @ error/ghostscript-private.h/InvokeGhostscriptDelegate/178.
```

**Reply by fmw42** [reply]
In fact, none of your files are readable when I download them.

**Reply by gmek5476** [reply]
Sorry i didn't get you

**Reply by fmw42** [reply]
I cannot open any of your images (PDF or PNG).  Something is wrong with them.

**Reply by gmek5476** [reply]
Sorry for inconvenience ,
Below are the other original assets Could you pls try with these
Source
![Ketchup-S](https://github.com/ImageMagick/ImageMagick/assets/126016862/be9b106e-48ed-4b88-bc92-01e40e777b42)
Target
![Ketchup-S1-cut3](https://github.com/ImageMagick/ImageMagick/assets/126016862/73146dba-482e-4f0d-b1c2-1bb7126a4303)

**Reply by fmw42** [reply]
Sorry, here is what I meant.

The size of your output is larger than the size of your input.  How have you changed the size?

**Reply by fmw42** [reply]
OK, new images. Let me look at your commands

**Reply by fmw42** [reply]
Is the second image, what you want? If not, what is wrong with it?

**Reply by gmek5476** [reply]
YEs second image is that i need ,But i cannot change the color and not sure how the polygon is drawing ,i got this command from google search i have a requirement to have projection based on angle.Means suppose from origin if the angle is 45 degrees my expectation to draw the Cut/Slant line at that angle from origin to detination.Could you explain  how polygon command is working and how we have to pass the co-ordinates.

**Reply by fmw42** [reply]
What angle do you want to use?

**Reply by gmek5476** [reply]
it is dynamic it can be 45degrees or 30 degrees etc.,

**Comment by fmw42** [comment]
First, in IM 7, use magick, not convert and not magick convert.  Second you should draw the black triangle on a white background rather than your white odd polygon on a black background.  Third use %[fx: ... ] computations for the point locations.  Since you have a triangle and want to draw the top right vertex use the height and the tan of the angle (actually 90 -angle) if angle is measured from horizontal to compute the x coordinate at y=0.

Unix syntax:

```
angle=45
magick 244506722-be9b106e-48ed-4b88-bc92-01e40e777b42.JPG \
\( +clone -fill white -colorize 100 -fill black -draw "polygon 0,0 0,%[fx:h-1] %[fx:h*tan(pi*(90-$angle)/180)],0" \) \
-alpha off -compose CopyOpacity -composite  \
result.png

```

<br>

Sorry, I do not use Windows. The command would be similar without the \ escapes on the parentheses and with ^ end of line marks in place of \.  Also double % to %% if in a bat file.  However, I do not know the Windows syntax for providing the angle variable as an argument.  You can hard code, however so that command for 45 degrees would be:

```
magick 244506722-be9b106e-48ed-4b88-bc92-01e40e777b42.JPG ^
( +clone -fill white -colorize 100 -fill black -draw "polygon 0,0 0,%[fx:h-1] %[fx:h*tan(pi*(90-45)/180)],0" ) ^
-alpha off -compose CopyOpacity -composite  ^
result.png
```

<br>

Replace 45 with whatever angle you want.


![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/2c40ea2c-3198-47ef-87ae-0420978146dc)

**Comment by fmw42** [comment]
If you want a different color other than transparent, just flatten against that background color

```
magick 244506722-be9b106e-48ed-4b88-bc92-01e40e777b42.JPG ^
( +clone -fill white -colorize 100 -fill black -draw "polygon 0,0 0,%[fx:h-1] %[fx:h*tan(pi*(90-45)/180)],0" ) ^
-alpha off -compose CopyOpacity -composite  ^
-background red -compose over -flatten ^
result.png
```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/c0fcb829-324b-4289-9c46-e540692961c7)

**Reply by gmek5476** [reply]
Hi Fred ,
Thanks a lot you are my saviour.One more request is that instead of having angle at (0,0) can we find the projection angle at certain vertex(x,y),Ofcourse it will not be a triangle shape.

**Reply by fmw42** [reply]
Please provide an image example or diagram of what you want. Sorry, I do not understand the request.

**Reply by gmek5476** [reply]
This is what iam expecting
<img width="750" alt="Result1" src="https://github.com/ImageMagick/ImageMagick/assets/126016862/4a7ac1e6-bc27-4dd0-abee-528fb7a66448">

**Reply by fmw42** [reply]
Sorry, still do not understand.  Draw an example of the way you want the result to look.

**Reply by gmek5476** [reply]
Sorry my bad here is the output iam expecting
![Ketchup-S - Copy (2)](https://github.com/ImageMagick/ImageMagick/assets/126016862/018f4f7f-fc07-49a0-a795-8ec5a6cd2a3d)

**Reply by fmw42** [reply]
So you want an X offset?  And you will provide that X-offset along the bottom of the image?  Is that correct? And the angle will be less than or equal to 90 deg?  Correct?

**Reply by gmek5476** [reply]
yes ,we will provide the x-offset and the deg is always less than or equal to  90

**Comment by fmw42** [comment]
Here is what I think you want.  

Unix Syntax for IM 7:

```
angle=45
xoff=200
magick 244506722-be9b106e-48ed-4b88-bc92-01e40e777b42.JPG \
\( +clone -fill white -colorize 100 -fill black -draw "polygon 0,0 0,%[fx:h-1] $xoff,%[fx:h-1] %[fx:h*tan(pi*(90-$angle)/180)],0" \) \
-alpha off -compose CopyOpacity -composite -compose over \
-background red -flatten \
result.png

```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/c21766a2-0de5-4bb2-8df1-4fddc7e4a902)

**Reply by gmek5476** [reply]
Great Thanks ,But the result is not correct when the angle is 60 .
Please find the command below
magick C:\Im-Test\Input\Ketchup-S.jpg ( +clone -fill white -colorize 100 -fill black -draw "polygon 0,0 0,%[fx:h-1] 750,%[fx:h-1] %[fx:h*tan(pi*(90-60)/180)],0" ) -alpha off -compose CopyOpacity -composite -compose over -background red -flatten C:\Im-Test\Output\cut-Out\Ketchup-S160.png
Please find the screenshot below
![Ketchup-S160](https://github.com/ImageMagick/ImageMagick/assets/126016862/01232df4-68a7-48c1-b0be-96f0c656fff8)

**Comment by fmw42** [comment]
Works fine for me

```
angle=60
xoff=200
magick 244506722-be9b106e-48ed-4b88-bc92-01e40e777b42.JPG \
\( +clone -fill white -colorize 100 -fill black -draw "polygon 0,0 0,%[fx:h-1] $xoff,%[fx:h-1] %[fx:h*tan(pi*(90-$angle)/180)],0" \) \
-alpha off -compose CopyOpacity -composite -compose over \
-background red -flatten \
result.png
```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/cee75b9e-5cc5-4de3-b691-756341ca32ac)

**Reply by fmw42** [reply]
You have at least one mistake, you have

`%[fx:htan(pi(90-60)/180)],0`

which should be

`%[fx:h*tan(pi(90-60)/180)],0`

**Reply by gmek5476** [reply]
Sorry "*" is there in my command ,HEre is the onw which i tried 
C:\ImageMagick-7>magick C:\Im-Test\Input\Ketchup-S.jpg ( +clone -fill white -colorize 100 -fill black -draw "polygon 0,0 0,%[fx:h-1] 750,%[fx:h-1] %[fx:h*tan(pi*(90-60)/180)],0" ) -alpha off -compose CopyOpacity -composite -compose over -background red -flatten C:\Im-Test\Output\cut-Out\Ketchup-S160.png.It's an issue with github comments it is removing "*" when i post the command.Pls Find the scree
<img width="736" alt="Capture" src="https://github.com/ImageMagick/ImageMagick/assets/126016862/b6d4c59b-80e3-4eec-a881-c84687c3ca66">
nshot below

**Reply by gmek5476** [reply]
Could you let me know the version of imagemagick? i tried with angle=60 and xoffset:750 but getting the response and shared earlier.
Woundering hw it is giving different for us

**Reply by fmw42** [reply]
Your command is wrong as I mentioned above.  Need h*tan not htan in your command

**Reply by fmw42** [reply]
I am using IM 7.1.1-10.  You have a typo as I mentioned above.

**Reply by fmw42** [reply]
Post your new image and I will test with that. But I do not see what could be wrong. It works for me as above.

**Reply by gmek5476** [reply]
IT's not a typo in comments it is taking off the "\*" symbol.PFA for command and output.I have tried with right command.
<img width="948" alt="Capture" src="https://github.com/ImageMagick/ImageMagick/assets/126016862/99ac6a7d-e1d6-4f77-8b7d-1eed57695887">

**Reply by fmw42** [reply]
If you are running in a .bat file, then you need to double the % to %%

**Reply by fmw42** [reply]
Please use the <> button after selecting the command to format it properly.  Do not post images of your commands.

**Reply by fmw42** [reply]
OK. I can reproduce it. I need to see why

**Reply by gmek5476** [reply]
cool Thanks Boss``

**Comment by fmw42** [comment]
I made a mistake. I forgot to add $xoff to the tan computation.  So `h*tan --> $xoff+h*tan`. 

So the correct command is:

```
angle=60
xoff=750
magick 244506722-be9b106e-48ed-4b88-bc92-01e40e777b42.JPG \
\( +clone -fill white -colorize 100 -fill black -draw "polygon 0,0 0,%[fx:h-1] $xoff,%[fx:h-1] %[fx:$xoff+h*tan(pi*(90-$angle)/180)],0" \) \
-alpha off -compose CopyOpacity -composite -compose over \
-background red -flatten \
result.png
```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/f0177151-496b-40af-be56-e278c0c386af)

**Reply by gmek5476** [reply]
Hi FRed ,

Need your help,Hope you remember we used below command to draw the polygon but there is one concern with the cross line. when we zoom the diagnol on Rightside the cross line appears like a steps. Is there any way we can correct it.
Here is the command iam using
magick C:\Im-Test\Input\pexels-photo-3028127-S.jpg   ( +clone -fill white -colorize 100 -fill black +antialias  -draw " polygon 0,0 0,%[fx:h-1] %[fx:(35/100)*w],%[fx:h-1] %[fx:((35/100)*w)+h*tan(pi*(90-70)/180)],0" ) -alpha off -compose CopyOpacity -composite -compose over -background yellow  -flatten C:\Im-Test\Output\cut-Out\issue12.jpg

Please find the source and target images 
![issue12](https://github.com/ImageMagick/ImageMagick/assets/126016862/ee71a226-f074-426e-9803-888de8dbe88d)
![pexels-photo-3028127-S](https://github.com/ImageMagick/ImageMagick/assets/126016862/4d2d3637-3e49-480f-9b50-89aef3100aef)
<img width="184" alt="Capture" src="https://github.com/ImageMagick/ImageMagick/assets/126016862/644dc5f3-882c-4529-baa3-eae204106fed">

**Comment by fmw42** [comment]
I do not understand the problem.  What is your complaint about the line drawn?

Fred
